### ISSUE TITLE
gradients through where when one of the branches is nan

### ISSUE THREAD

**Issue by rajeshpoddar** [tinygrad/tinygrad#12409]
What is tinygrad's plan to deal with gradients of functions defined with a Tensor.where operation in which the branch not taken is nan but the gradient still comes out to nan due 0*nan resulting in nan. The issue is present in jax as well and is described in detail here - https://github.com/jax-ml/jax/issues/1052#issuecomment-514083352.

It can be reproduced in tinygrad with something like
`from tinygrad import Tensor

x = Tensor([0], dtype="float32", requires_grad=True)
y = Tensor.where(x == 0, 0, x * x.sqrt())
y.sum().backward()
print(x.grad.tolist())`

As they mention, the right solution which "is pretty heavyweight", "is something like tracking a symbolic zero mask potentially through the entire backward pass of differentiation".

**Comment by chenyuxyz** [comment]
we don't have a clear spec for nan, so the goal would be if the correct answer is unambiguously not nan, we should have that. otherwise it's undefined

interesting if you replace `x * x.sqrt()` with `x ** 1.5` it returns 0.0, as pow has a special branch to deal with nan case
